kernel/include/circular_queue_kern.h

```diff
@@ -0,0 +1,128 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+/** @brief Circular queue implementation for a producer, consumer based setup.
+ *
+ *  Specifically this circular queue implementation is used for passing event
+ *  information from the kernel to the user.
+ *
+ *
+ *  For safety this queue on an error should log the error and reset to a known
+ *  safe state possibly dropping all data in held within it.
+ *
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include <sys/lock.h>
+
+#include "feeds.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Circular queue data structure.
+typedef struct osquery_cqueue {
+  uint8_t *buffer;
+  size_t size;
+  size_t produced;
+  uint8_t *write;
+  uint8_t *max_read;
+  uint8_t *read;
+  int no_reserved;
+  int empty;
+  int dropped;
+  int initialized;
+
+  lck_grp_attr_t *lck_grp_attr;
+  lck_grp_t *lck_grp;
+  lck_attr_t *lck_attr;
+  lck_spin_t *lck;
+
+  lck_mtx_t *mtx;  // For msleep.
+} osquery_cqueue_t;
```

Comment:
You shouldn't use *_t type names (they are reserved under POSIX).
http://www.gnu.org/software/libc/manual/html_node/Reserved-Names.html